File: src/meshioplusplus/_data_condition.py
"""Value conditioning for data arrays: clamp, normalize, standardize.

A dependency-free mesh *operation* (not a file format). The geometry is never
modified; only the values of the selected arrays change.

- ``clamp`` maps each value to ``min(max(x, lo), hi)``.
- ``normalize`` maps the array's own ``[min, max]`` linearly onto ``[lo, hi]``
  (``[0, 1]`` by default).
- ``standardize`` maps to zero mean and unit standard deviation (population
  standard deviation, divisor N).

``scope="component"`` (the default) conditions each trailing component on its
own statistics; ``scope="magnitude"`` uses each row's Euclidean magnitude and
rescales whole rows, preserving direction.

For ``cell_data`` the statistics are computed **jointly over all cell blocks**,
then one transform is applied to every block.

Non-finite values are always excluded from the statistics; ``nan_policy``
decides what reaches the output. See :mod:`meshioplusplus._data_common`.

Public API:
    data_condition
"""

from __future__ import annotations

import copy
import warnings

import numpy as np

from ._data_common import location_map, normalize_location, require_key


def _components(array):
    a = np.asarray(array)
    return 1 if a.ndim < 2 else int(np.prod(a.shape[1:]))


def _finite_stats(values, ncomp):
    """Per-component (min, max, mean, std) over the finite values."""
    flat = np.asarray(values, dtype=float).reshape(-1, ncomp)
    mask = np.isfinite(flat)
    counts = mask.sum(axis=0)
    safe = np.where(mask, flat, np.nan)
    with warnings.catch_warnings():
        warnings.simplefilter("ignore", RuntimeWarning)
        mins = np.nanmin(safe, axis=0)
        maxs = np.nanmax(safe, axis=0)
        means = np.nanmean(safe, axis=0)
        stds = np.nanstd(safe, axis=0)
    mins = np.where(counts > 0, mins, np.nan)
    maxs = np.where(counts > 0, maxs, np.nan)
    means = np.where(counts > 0, means, np.nan)
    stds = np.where(counts > 0, stds, np.nan)
    return mins, maxs, means, stds, counts


def _magnitudes(values, ncomp):
    """Row magnitudes; NaN for any row containing a non-finite component."""
    flat = np.asarray(values, dtype=float).reshape(-1, ncomp)
    finite = np.isfinite(flat).all(axis=1)
    mags = np.sqrt((np.where(np.isfinite(flat), flat, 0.0) ** 2).sum(axis=1))
    return np.where(finite, mags, np.nan)


def _build(mode, lo, hi, mins, maxs, means, stds, counts, name):
    """Return per-component (scale, offset) for the affine modes."""
    if mode == "normalize":
        span = maxs - mins
        ok = (counts > 0) & np.isfinite(span) & (span > 0)
        if not np.all(ok):
            warnings.warn(
                f"meshio++: data_condition: '{name}' has no value range to "
                "normalize (constant or non-finite); filling with the target "
                "lower bound",
                RuntimeWarning,
                stacklevel=2,
            )
        scale = np.where(ok, (hi - lo) / np.where(ok, span, 1.0), 0.0)
        offset = np.where(ok, lo - scale * mins, lo)
        return scale, offset
    if mode == "standardize":
        ok = (counts > 0) & np.isfinite(stds) & (stds > 0)
        if not np.all(ok):
            warnings.warn(
                f"meshio++: data_condition: '{name}' has zero standard "
                "deviation; filling with 0",
                RuntimeWarning,
                stacklevel=2,
            )
        scale = np.where(ok, 1.0 / np.where(ok, stds, 1.0), 0.0)
        offset = np.where(ok, -means * scale, 0.0)
        return scale, offset
    raise ValueError(
        f"meshio++: unknown conditioning mode '{mode}' "
        "(expected 'clamp', 'normalize' or 'standardize')"
    )


def _apply_nan_policy(values, nan_policy, nan_replacement, name):
    if nan_policy == "ignore":
        return values
    bad = ~np.isfinite(values)
    if not bad.any():
        return values
    if nan_policy == "replace":
        out = values.copy()
        out[bad] = nan_replacement
        return out
    if nan_policy == "fail":
        idx = int(np.argmax(bad.ravel()))
        raise ValueError(
            f"meshio++: data_condition: non-finite value at index {idx} of '{name}'"
        )
    raise ValueError(
        f"meshio++: unknown NaN policy '{nan_policy}' "
        "(expected 'ignore', 'replace' or 'fail')"
    )


def _condition_py(
    mesh,
    location,
    names,
    mode,
    scope,
    lo,
    hi,
    nan_policy,
    nan_replacement,
    suffix,
    preserve_dtype,
):
    """Pure-Python reference for :func:`data_condition`."""
    loc = normalize_location(location)
    if mode in ("clamp", "normalize") and lo > hi:
        raise ValueError(
            f"meshio++: data_condition: bounds are inverted (lo={lo} > hi={hi})"
        )
    source = location_map(mesh, loc)
    selected = list(names) if names else sorted(source.keys())
    for name in selected:
        require_key(mesh, loc, name)

    out = copy.deepcopy(mesh)
    magnitude = scope == "magnitude"
    if scope not in ("component", "magnitude"):
        raise ValueError(
            f"meshio++: unknown conditioning scope '{scope}' "
            "(expected 'component' or 'magnitude')"
        )

    for name in selected:
        target = name + suffix
        if loc == "cell_data":
            blocks = source[name]
            if len(blocks) != len(mesh.cells):
                raise ValueError(
                    f"meshio++: data_condition: cell_data '{name}' has "
                    f"{len(blocks)} block(s) but the mesh has "
                    f"{len(mesh.cells)} cell block(s)"
                )
            arrays = [np.asarray(b) for b in blocks]
        else:
            arrays = [np.asarray(source[name])]

        ncomp = _components(arrays[0])
        pooled = (
            np.concatenate([_magnitudes(a, ncomp) for a in arrays]).reshape(-1, 1)
            if magnitude
            else np.concatenate(
                [np.asarray(a, dtype=float).reshape(-1, ncomp) for a in arrays]
            )
        )
        stat_comp = 1 if magnitude else ncomp
        mins, maxs, means, stds, counts = _finite_stats(pooled, stat_comp)
        if mode != "clamp":
            scale, offset = _build(mode, lo, hi, mins, maxs, means, stds, counts, name)

        results = []
        for a in arrays:
            rows = a.shape[0] if a.ndim else 0
            flat = np.asarray(a, dtype=float).reshape(rows, ncomp)
            if magnitude:
                mags = _magnitudes(a, ncomp)
                if mode == "clamp":
                    target_mag = np.clip(mags, lo, hi)
                else:
                    target_mag = scale[0] * mags + offset[0]
                with np.errstate(invalid="ignore", divide="ignore"):
                    factor = np.where(
                        mags > 0, target_mag / np.where(mags > 0, mags, 1.0), 0.0
                    )
                res = flat * factor.reshape(-1, 1)
                res = np.where(np.isfinite(mags).reshape(-1, 1), res, flat)
            elif mode == "clamp":
                res = np.where(np.isfinite(flat), np.clip(flat, lo, hi), flat)
            else:
                res = np.where(np.isfinite(flat), flat * scale + offset, flat)

            res = _apply_nan_policy(res, nan_policy, nan_replacement, name)
            if mode == "clamp" and preserve_dtype:
                res = res.astype(np.asarray(a).dtype)
            shaped = (
                res.reshape(np.asarray(a).shape) if ncomp > 1 else res.reshape(rows)
            )
            results.append(shaped)

        if loc == "cell_data":
            out.cell_data[target] = results
        elif loc == "point_data":
            out.point_data[target] = results[0]
        else:
            out.field_data[target] = results[0]
    return out


def data_condition(
    mesh,
    location: str = "point",
    keys=None,
    mode: str = "clamp",
    scope: str = "component",
    lo: float = 0.0,
    hi: float = 1.0,
    nan_policy: str = "ignore",
    nan_replacement: float = 0.0,
    suffix: str = "",
    preserve_dtype: bool = True,
):
    """Condition the values of the selected data arrays.

    Args:
        mesh: the source mesh (unmodified).
        location: ``"point"``, ``"cell"`` or ``"field"``.
        keys: names to condition; ``None`` means every array at ``location``.
        mode: ``"clamp"``, ``"normalize"`` or ``"standardize"``.
        scope: ``"component"`` or ``"magnitude"``.
        lo: clamp lower bound, or normalize target lower bound.
        hi: clamp upper bound, or normalize target upper bound.
        nan_policy: ``"ignore"``, ``"replace"`` or ``"fail"``.
        nan_replacement: value used when ``nan_policy`` is ``"replace"``.
        suffix: empty replaces the array in place; otherwise the result is
            stored as ``name + suffix``.
        preserve_dtype: ``clamp`` only -- keep the input dtype. ``normalize``
            and ``standardize`` always produce ``float64``.

    Returns:
        A new mesh with the conditioned arrays.
    """
    loc = normalize_location(location)
    names = list(keys) if keys else []
    args = (
        loc,
        names,
        mode,
        scope,
        lo,
        hi,
        nan_policy,
        nan_replacement,
        suffix,
        preserve_dtype,
    )
    try:
        from . import _core
    except Exception:
        return _condition_py(mesh, *args)
    try:
        out = _core.data_condition(
            mesh,
            loc,
            names,
            mode,
            scope,
            lo,
            hi,
            nan_policy,
            nan_replacement,
            suffix,
            preserve_dtype,
        )
    except ValueError:
        raise
    except Exception:
        return _condition_py(mesh, *args)
    for attr in ("point_sets", "cell_sets"):
        value = getattr(mesh, attr, None)
        if value:
            setattr(out, attr, copy.deepcopy(value))
    return out

File: src/cpp/src/operations/data_condition.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//
// Value conditioning for data arrays: clamp / normalize / standardize, either
// per component or by row magnitude. Statistics come from the shared
// detail::FiniteStats chunked reducer (deterministic: parallel chunks, serial
// combine) and, for cell_data, are gathered across ALL blocks before a single
// transform is applied to each -- normalizing one block against its own
// extremes would make the pieces mutually incomparable.
//
// Geometry is never modified. See operations/data_condition.hpp for the
// contract.

// System includes
#include <cmath>
#include <cstddef>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

// Project includes
#include "meshioplusplus/operations/data_condition.hpp"
#include "meshioplusplus/detail/data_ops.hpp"
#include "meshioplusplus/detail/value_io.hpp"
#include "meshioplusplus/log.hpp"
#include "meshioplusplus/ndarray.hpp"
#include "meshioplusplus/operations/data_common.hpp"
#include "meshioplusplus/parallel.hpp"

namespace meshioplusplus {

namespace {

/// The affine map `y = scale * x + offset` a mode/scope resolves to, per
/// component (or a single entry when conditioning by magnitude).
struct DcondTransform {
    double mScale = 1.0;
    double mOffset = 0.0;
    bool mClamp = false;  ///< Clamp mode: saturate into [mLo, mHi] instead.
    double mLo = 0.0;
    double mHi = 1.0;
};

/// Applies the NaN policy to one produced value.
double dcond_apply_nan_policy(double v, const DataConditionOptions& rOpts, const std::string& rName,
                              std::size_t index) {
    if (std::isfinite(v))
        return v;
    switch (rOpts.nan_policy) {
        case NanPolicy::Ignore:
            return v;
        case NanPolicy::Replace:
            return rOpts.nan_replacement;
        case NanPolicy::Fail:
            throw std::invalid_argument("meshio++: data_condition: non-finite value at index " +
                                        std::to_string(index) + " of '" + rName + "'");
    }
    return v;
}

/// Builds the transform for one component from its statistics.
DcondTransform dcond_build(const detail::FiniteStats& rStats, const DataConditionOptions& rOpts,
                           const std::string& rName, bool& rWarnedConstant) {
    DcondTransform t;
    switch (rOpts.mode) {
        case ConditionMode::Clamp:
            t.mClamp = true;
            t.mLo = rOpts.lo;
            t.mHi = rOpts.hi;
            return t;
        case ConditionMode::Normalize: {
            const double lo = rStats.Min();
            const double hi = rStats.Max();
            if (rStats.mNumFinite == 0 || !(hi > lo)) {
                // Constant (or entirely non-finite) array: everything maps to
                // the target lower bound. Warn once per array.
                if (!rWarnedConstant) {
                    rWarnedConstant = true;
                    log::warn(
                        "data_condition: '{}' has no value range to normalize (constant or "
                        "non-finite); filling with the target lower bound",
                        rName);
                }
                t.mScale = 0.0;
                t.mOffset = rOpts.lo;
                return t;
            }
            t.mScale = (rOpts.hi - rOpts.lo) / (hi - lo);
            t.mOffset = rOpts.lo - t.mScale * lo;
            return t;
        }
        case ConditionMode::Standardize: {
            const double sd = rStats.StdDev();
            const double mean = rStats.Mean();
            if (rStats.mNumFinite == 0 || !(sd > 0.0)) {
                if (!rWarnedConstant) {
                    rWarnedConstant = true;
                    log::warn("data_condition: '{}' has zero standard deviation; filling with 0",
                              rName);
                }
                t.mScale = 0.0;
                t.mOffset = 0.0;
                return t;
            }
            t.mScale = 1.0 / sd;
            t.mOffset = -mean / sd;
            return t;
        }
    }
    return t;
}

/// Applies one transform to a single value.
double dcond_apply(const DcondTransform& rT, double v) {
    if (!std::isfinite(v))
        return v;  // policy handled separately; never fold NaN into arithmetic
    if (rT.mClamp)
        return v < rT.mLo ? rT.mLo : (v > rT.mHi ? rT.mHi : v);
    return rT.mScale * v + rT.mOffset;
}

/// Reduces the row magnitudes of an array into one accumulator.
void dcond_accumulate_magnitude(const NDArray& rArray, std::size_t ncomp,
                                detail::FiniteStats& rStats) {
    const std::size_t total = rArray.Size();
    if (total == 0 || ncomp == 0)
        return;
    const std::size_t nrows = total / ncomp;
    const std::size_t grain = 4096;
    const std::size_t nchunks = (nrows + grain - 1) / grain;
    std::vector<detail::FiniteStats> partial(nchunks);
    parallel_for(
        nchunks,
        [&](std::size_t ci) {
            detail::FiniteStats local;
            const std::size_t begin = ci * grain;
            const std::size_t end = std::min(begin + grain, nrows);
            for (std::size_t r = begin; r < end; ++r) {
                double acc = 0.0;
                bool finite = true;
                for (std::size_t k = 0; k < ncomp; ++k) {
                    const double v = detail::read_double(rArray, r * ncomp + k);
                    if (!std::isfinite(v)) {
                        finite = false;
                        break;
                    }
                    acc += v * v;
                }
                local.Add(finite ? std::sqrt(acc) : std::nan(""));
            }
            partial[ci] = local;
        },
        1);
    for (const detail::FiniteStats& p : partial)
        rStats.Merge(p);
}

/// The dtype the output array should carry.
DType dcond_output_dtype(const NDArray& rSource, const DataConditionOptions& rOpts) {
    if (rOpts.mode == ConditionMode::Clamp && rOpts.preserve_dtype)
        return rSource.Dtype();
    return DType::Float64;
}

/// Conditions one array (component scope).
NDArray dcond_transform_component(const NDArray& rSource, std::size_t ncomp,
                                  const std::vector<DcondTransform>& rTransforms,
                                  const DataConditionOptions& rOpts, const std::string& rName) {
    NDArray dst = NDArray::Uninit(dcond_output_dtype(rSource, rOpts), rSource.Shape());
    const std::size_t total = rSource.Size();
    // Elementwise and bandwidth-bound.
    parallel_for_bw(total, [&](std::size_t i) {
        const double v = detail::read_double(rSource, i);
        detail::write_double(dst, i, dcond_apply(rTransforms[i % ncomp], v));
    });
    // The NaN policy runs serially so a `Fail` throw is deterministic.
    if (rOpts.nan_policy != NanPolicy::Ignore)
        for (std::size_t i = 0; i < total; ++i) {
            const double v = detail::read_double(dst, i);
            detail::write_double(dst, i, dcond_apply_nan_policy(v, rOpts, rName, i));
        }
    return dst;
}

/// Conditions one array (magnitude scope): each row is rescaled by the factor
/// that its magnitude would have received, so the direction is preserved.
NDArray dcond_transform_magnitude(const NDArray& rSource, std::size_t ncomp,
                                  const DcondTransform& rT, const DataConditionOptions& rOpts,
                                  const std::string& rName) {
    NDArray dst = NDArray::Uninit(dcond_output_dtype(rSource, rOpts), rSource.Shape());
    const std::size_t total = rSource.Size();
    const std::size_t nrows = ncomp > 0 ? total / ncomp : 0;
    parallel_for_bw(nrows, [&](std::size_t r) {
        double acc = 0.0;
        bool finite = true;
        for (std::size_t k = 0; k < ncomp; ++k) {
            const double v = detail::read_double(rSource, r * ncomp + k);
            if (!std::isfinite(v)) {
                finite = false;
                break;
            }
            acc += v * v;
        }
        const double mag = finite ? std::sqrt(acc) : std::nan("");
        // Scale the whole row so its magnitude becomes the conditioned value.
        // A zero-magnitude row has no direction to preserve and stays zero.
        const double target = dcond_apply(rT, mag);
        const double factor = (finite && mag > 0.0) ? target / mag : (finite ? 0.0 : mag);
        for (std::size_t k = 0; k < ncomp; ++k) {
            const double v = detail::read_double(rSource, r * ncomp + k);
            detail::write_double(dst, r * ncomp + k, finite ? v * factor : v);
        }
    });
    if (rOpts.nan_policy != NanPolicy::Ignore)
        for (std::size_t i = 0; i < total; ++i) {
            const double v = detail::read_double(dst, i);
            detail::write_double(dst, i, dcond_apply_nan_policy(v, rOpts, rName, i));
        }
    return dst;
}

/// Resolves the requested names, validating each.
std::vector<std::string> dcond_resolve_names(const Mesh& rMesh, const DataConditionOptions& rOpts) {
    if (rOpts.names.empty())
        return data_names(rMesh, rOpts.location);
    for (const std::string& n : rOpts.names)
        if (!data_has(rMesh, rOpts.location, n))
            throw std::invalid_argument(data_unknown_key_message(rMesh, rOpts.location, n));
    return rOpts.names;
}

}  // namespace

ConditionMode condition_mode_from_name(const std::string& rName) {
    if (rName == "clamp")
        return ConditionMode::Clamp;
    if (rName == "normalize" || rName == "rescale")
        return ConditionMode::Normalize;
    if (rName == "standardize" || rName == "zscore")
        return ConditionMode::Standardize;
    throw std::invalid_argument("meshio++: unknown conditioning mode '" + rName +
                                "' (expected 'clamp', 'normalize' or 'standardize')");
}

ConditionScope condition_scope_from_name(const std::string& rName) {
    if (rName == "component")
        return ConditionScope::Component;
    if (rName == "magnitude")
        return ConditionScope::Magnitude;
    throw std::invalid_argument("meshio++: unknown conditioning scope '" + rName +
                                "' (expected 'component' or 'magnitude')");
}

Mesh data_condition(const Mesh& rMesh, const DataConditionOptions& rOpts) {
    if (rOpts.mode == ConditionMode::Clamp && rOpts.lo > rOpts.hi)
        throw std::invalid_argument(
            "meshio++: data_condition: clamp bounds are inverted (lo=" + std::to_string(rOpts.lo) +
            " > hi=" + std::to_string(rOpts.hi) + ")");
    if (rOpts.mode == ConditionMode::Normalize && rOpts.lo > rOpts.hi)
        throw std::invalid_argument(
            "meshio++: data_condition: normalize target range is inverted (lo=" +
            std::to_string(rOpts.lo) + " > hi=" + std::to_string(rOpts.hi) + ")");

    const std::vector<std::string> names = dcond_resolve_names(rMesh, rOpts);
    Mesh out = detail::clone_mesh(rMesh);
    const bool magnitude = rOpts.scope == ConditionScope::Magnitude;

    for (const std::string& name : names) {
        const std::string target = name + rOpts.suffix;

        if (rOpts.location == DataLocation::Cell) {
            const std::size_t nblocks = rMesh.NumCellBlocks();
            if (rMesh.CellDataNumBlocks(name) != nblocks)
                throw std::invalid_argument(
                    "meshio++: data_condition: cell_data '" + name + "' has " +
                    std::to_string(rMesh.CellDataNumBlocks(name)) + " block(s) but the mesh has " +
                    std::to_string(nblocks) + " cell block(s)");
            if (nblocks == 0)
                continue;
            const std::size_t ncomp = data_num_components(rMesh.CellData(name, 0));

            // Statistics across ALL blocks first.
            std::vector<detail::FiniteStats> stats;
            detail::FiniteStats magstats;
            for (std::size_t b = 0; b < nblocks; ++b) {
                const NDArray& src = rMesh.CellData(name, b);
                if (data_num_components(src) != ncomp)
                    throw std::invalid_argument(
                        "meshio++: data_condition: cell_data '" + name +
                        "' has inconsistent component counts across blocks");
                if (magnitude)
                    dcond_accumulate_magnitude(src, ncomp, magstats);
                else
                    detail::accumulate_stats(src, ncomp, stats);
            }

            bool warned = false;
            std::vector<DcondTransform> transforms;
            DcondTransform magt;
            if (magnitude) {
                magt = dcond_build(magstats, rOpts, name, warned);
            } else {
                transforms.reserve(ncomp);
                for (std::size_t k = 0; k < ncomp; ++k)
                    transforms.push_back(dcond_build(stats[k], rOpts, name, warned));
            }

            std::vector<NDArray> blocks;
            blocks.reserve(nblocks);
            for (std::size_t b = 0; b < nblocks; ++b) {
                const NDArray& src = rMesh.CellData(name, b);
                blocks.push_back(
                    magnitude ? dcond_transform_magnitude(src, ncomp, magt, rOpts, name)
                              : dcond_transform_component(src, ncomp, transforms, rOpts, name));
            }
            out.AddCellData(target, std::move(blocks));
            continue;
        }

        const NDArray& src =
            rOpts.location == DataLocation::Point ? rMesh.PointData(name) : rMesh.FieldData(name);
        const std::size_t ncomp = data_num_components(src);
        bool warned = false;
        NDArray dst;
        if (magnitude) {
            detail::FiniteStats magstats;
            dcond_accumulate_magnitude(src, ncomp, magstats);
            const DcondTransform t = dcond_build(magstats, rOpts, name, warned);
            dst = dcond_transform_magnitude(src, ncomp, t, rOpts, name);
        } else {
            std::vector<detail::FiniteStats> stats;
            detail::accumulate_stats(src, ncomp, stats);
            std::vector<DcondTransform> transforms;
            transforms.reserve(ncomp);
            for (std::size_t k = 0; k < ncomp; ++k)
                transforms.push_back(dcond_build(stats[k], rOpts, name, warned));
            dst = dcond_transform_component(src, ncomp, transforms, rOpts, name);
        }
        if (rOpts.location == DataLocation::Point)
            out.AddPointData(target, std::move(dst));
        else
            out.AddFieldData(target, std::move(dst));
    }
    return out;
}

}  // namespace meshioplusplus
